src/main/kotlin/uk/gov/justice/digital/hmpps/nomisvisitsmappingservice/service/LocationMappingService.kt

```diff
@@ -0,0 +1,138 @@
+package uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.service
+
+import com.microsoft.applicationinsights.TelemetryClient
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.toList
+import org.slf4j.LoggerFactory
+import org.springframework.data.domain.Page
+import org.springframework.data.domain.PageImpl
+import org.springframework.data.domain.Pageable
+import org.springframework.stereotype.Service
+import org.springframework.transaction.annotation.Transactional
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.config.DuplicateMappingException
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.data.LocationMappingDto
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.jpa.LocationMapping
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.jpa.LocationMappingType
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.jpa.LocationMappingType.MIGRATED
+import uk.gov.justice.digital.hmpps.nomisvisitsmappingservice.jpa.repository.LocationMappingRepository
+
+@Service
+@Transactional(readOnly = true)
+class LocationMappingService(
+  private val locationMappingRepository: LocationMappingRepository,
+  private val telemetryClient: TelemetryClient,
+) {
+  private companion object {
+    private val log = LoggerFactory.getLogger(this::class.java)
+  }
+
+  fun alreadyExistsMessage(
+    duplicateMapping: LocationMappingDto,
+    existingMapping: LocationMappingDto,
+  ) =
+    "Location mapping already exists. \nExisting mapping: $existingMapping\nDuplicate mapping: $duplicateMapping"
+
+  @Transactional
+  suspend fun createMapping(createMappingRequest: LocationMappingDto) =
+    with(createMappingRequest) {
+      log.debug("creating location {}", createMappingRequest)
+      locationMappingRepository.findById(dpsLocationId)?.run {
+        if (this@run.nomisLocationId == this@with.nomisLocationId) {
+          log.debug(
+            "Not creating. All OK: " +
```

Comment:
bit naff adding strings together - should really use `{}` with slf4j

Reply:
Done for all occurrances